src/luascript.cpp

```diff
@@ -8523,27 +8526,41 @@ int LuaScriptInterface::luaPlayerShowTextDialog(lua_State* L)
 int LuaScriptInterface::luaPlayerSendTextMessage(lua_State* L)
 {
 	// player:sendTextMessage(type, text[, position, primaryValue = 0, primaryColor = TEXTCOLOR_NONE[, secondaryValue = 0, secondaryColor = TEXTCOLOR_NONE]])
-	int parameters = lua_gettop(L);
+	// player:sendTextMessage(type, text, channelId)
 
-	TextMessage message(getNumber<MessageClasses>(L, 2), getString(L, 3));
-	if (parameters >= 6) {
-		message.position = getPosition(L, 4);
-		message.primary.value = getNumber<int32_t>(L, 5);
-		message.primary.color = getNumber<TextColor_t>(L, 6);
+	Player* player = getUserdata<Player>(L, 1);
+	if (!player) {
+		lua_pushnil(L);
+		return 1;
 	}
 
-	if (parameters >= 8) {
-		message.secondary.value = getNumber<int32_t>(L, 7);
-		message.secondary.color = getNumber<TextColor_t>(L, 8);
-	}
+	int parameters = lua_gettop(L);
 
-	Player* player = getUserdata<Player>(L, 1);
-	if (player) {
-		player->sendTextMessage(message);
-		pushBoolean(L, true);
+	TextMessage message(getNumber<MessageClasses>(L, 2), getString(L, 3));
+	if (parameters == 4) {
+		uint16_t channelId = getNumber<uint16_t>(L, 4);
+		ChatChannel* channel = g_chat->getChannel(*player, channelId);
+		if (!channel || !channel->hasUser(*player)) {
+			pushBoolean(L, false);
+			return 1;
+		}
+		message.channelId = channelId;
 	} else {
-		lua_pushnil(L);
+		if (parameters >= 6) {
+			message.position = getPosition(L, 4);
+			message.primary.value = getNumber<int32_t>(L, 5);
+			message.primary.color = getNumber<TextColor_t>(L, 6);
+		}
+
+		if (parameters >= 8) {
+			message.secondary.value = getNumber<int32_t>(L, 7);
+			message.secondary.color = getNumber<TextColor_t>(L, 8);
+		}
```

Comment:
Are you sure? Indeed you will have to swap them, so that `>= 8` is checked before `>= 6`, and each one on its own `else if` branch. The `else` branch is when not enough parameters were supplied and you can push nil and return 1 directly.